### ISSUE TITLE
1k sounds training, best params? xl-sft inference?

### ISSUE THREAD

**Issue by Katehuuh** [koda-dernet/Side-Step#57]
What are the recommended params for 1.3k pop tracks (well-labeled lyrics + description)? Which model? How many epochs?

I used the xl-sft largest checkpoint hoping for best quality, rank 64, alpha 128, adamw8bit, chunk-duration 60, trained [100 epochs](https://huggingface.co/WeReCooking/ACE-Step-xl-sft-LoRA-1374files) and hit a plateau, best MA5 0.8371. But the audio just loops repeatedly. Does that mean it's a bad loss? Undertrained?

<details>
<summary>Training command + Inference params</summary>

```
uv run sidestep --yes --plain train ^
    --checkpoint-dir "%CHECKPOINT_DIR%" ^
    --model xl-sft ^
    --dataset-dir "%TENSOR_DIR%" ^
    --output-dir "%OUTPUT_DIR%" ^
    --adapter lora ^
    --rank 64 --alpha 128 --dropout 0.1 ^
    --attention-type both ^
    --target-mlp ^
    --epochs 200 ^
    --lr 3e-4 ^
    --batch-size 2 ^
    --gradient-accumulation 4 ^
    --warmup-steps 100 ^
    --optimizer-type adamw8bit ^
    --scheduler-type cosine ^
    --gradient-checkpointing ^
    --chunk-duration 60 ^
    --save-every 50 ^
    --save-best-after 50 ^
    --device cuda:0 ^
    --precision auto ^
    %RESUME_FLAG%
```

```python
# infer.py edit for xl-sft

model.generate_audio(
    text_hidden_states=text_hs,           # [1, 256, 1024]
    text_attention_mask=text_mask,
    lyric_hidden_states=lyric_hs,          # [1, 512, 1024] from embed_tokens
    lyric_attention_mask=lyric_mask,
    refer_audio_acoustic_hidden_states_packed=torch.zeros(1, 1, 64),
    refer_audio_order_mask=torch.zeros(1, dtype=long),
    src_latents=lm_context_latents,        # [1, 1500, 64]
    chunk_masks=torch.ones(1, 1500, 64),
    is_covers=torch.ones(1, dtype=long),
    silence_latent=silence,                # [1, 1500, 64]
    seed=42,
    fix_nfe=50,
    shift=1.0,                             # default for sft (turbo=3.0)
)
```

LM: `temperature=1.0, top_p=0.95, top_k=0` (defaults).

## Stack:

- Model: `acestep-v15-xl-sft` (4.99B `AceStepConditionGenerationModel`)
- VAE: `AutoencoderOobleck` (decoder_input_channels=64, 48kHz stereo)
- Text encoder: `Qwen3-Embedding-0.6B`
- LM: `acestep-5Hz-lm-1.7B` (Qwen3 1.7B)

## Pipeline:

1. LM (Qwen3 1.7B) → audio codes from caption + lyrics
2. `text_encoder(ids).last_hidden_state` for caption (max_length=256)
3. `text_encoder.embed_tokens(ids)` for lyrics (max_length=512)
4. Detokenize LM codes → context latents
5. `model.generate_audio(...)` with `is_covers=1` when LM codes provided
6. VAE decode → 48kHz stereo

</details>

Is my edited inference the issue? The official [infer.py](https://github.com/ace-step/ACE-Step/blob/main/infer.py) only supports the non-xl variant. Does Side-Step always do inference via CLI? I avoid using ComfyUI.

**Comment by koda-dernet** [comment]
I would genuinely not use a Low Rank Adapter for training 1.3k samples. Adapters (most of them) are additive weights, best suited for samples in the sub hundred, much less a LoRA, which is the most primitive out of the adapters i offer. 

However if you must still train 1.3k tracks (for... some reason)

Change your adapter type to LoHA
64 is too low of a rank for a sample count that big, you might want to ramp it up a bit.
If you don't have the compute, do not use batch size, use 8 gradient accumulation steps instead
Stop chunking down to 60s, do 90s (or more) instead
If you have the compute, use fp32 precision, it will lower back down to sdp attention (which is not as fast and not as efficient as flash attention) but the results are worth it
Train on base, trust me, train on base

**Comment by Katehuuh** [comment]
> If you don't have the compute

On 5090

> I would genuinely not use a Low Rank Adapter for training 1.3k samples. Adapters (most of them) are additive weights, best suited for samples in the sub hundred, much less a LoRA, which is the most primitive out of the adapters i offer.

Then, would full fine tune help? Gradients + Model weights fp16 would fit.

> Change your adapter type to LoHA

[Ace-step/ACE-Step](https://github.com/ace-step/ACE-Step) Gradio does not support LoHA/LyCORIS, only [billwuhao/ComfyUI_ACE-Step](https://github.com/billwuhao/ComfyUI_ACE-Step) for inference? Or could I use a LoRA/LoHA merger-to-base script if this could be added in Side-Step? (Official ACE-Step doesn't include one, i can make my own but can't guarantee it works as expected.)

In the meantime, i will run:

```cmd
  uv run sidestep --yes --plain train `
      -d "../tmp/xl_tensors" `
      --checkpoint-dir "../tmp/checkpoints" `
      --model xl-base `
      --adapter loha `
      --loha-linear-dim 128 `
      --loha-linear-alpha 256 `
      --target-mlp `
      --batch-size 2 `
      --gradient-accumulation 4 `
      --chunk-duration 120 `
      --precision fp32 `
      --lr 5e-5 `
      --warmup-steps 300 `
      --optimizer-type adamw8bit `
      --scheduler-type cosine `
      --gradient-checkpointing `
      --gradient-checkpointing-ratio 0.5 `
      --epochs 100 `
      --save-every 5 `
      --output-dir "../tmp/train_runs/loha_xl_base_v4_100ep"
```
(higher dim 256/alpha 512 diverged the loss & used 27 GB)

**Comment by Katehuuh** [comment]
@koda-dernet I got it working, good enuf but expected... better sounding:
- 100 epochs: Mean 0.8265 → 0.7955, MA5 → 0.7955 for ~48h.
- <details><summary>I attempted to merge the LoHA into the base so I could then use <a href="https://github.com/ServeurpersoCom/acestep.cpp">acestep.cpp GGUF</a> with <code>merge_loha.py</code>, but wasn't able to get clear sound; <a href="https://github.com/ace-step/ACE-Step-1.5/blob/main/cli.py">ACE-Step-1.5/cli.py</a> also doesn't support LyCORIS, so I landed on the Python API approach via <code>AceStepHandler.add_lora()</code>. You can hear the result <a href="https://huggingface.co/Nekochu/ACE-Step-xl-base-pop-electro-lora">here</a>:</summary>

  ```python
  """ **Broken** - Merge a Side-Step LoHA adapter into the base xl-base weights.

  Output: a complete xl-base-style checkpoint with LoHA baked into the weights.
  Loads with ANY ACE-Step inference path (no adapter wrangling).

  Usage:
      # From inside a Side-Step clone (https://github.com/koda-dernet/Side-Step)
      uv run python merge_loha.py \\
          --base-ckpt-dir ./checkpoints \\
          --variant xl-base \\
          --adapter-dir ./output/my_lora \\
          --out ./checkpoints/acestep-v15-xl-base-merged \\
          --device cpu
  """
  import argparse
  import shutil
  import sys
  from pathlib import Path

  import torch
  from safetensors.torch import load_file as load_st

  from sidestep_engine.core.configs import LoHAConfigV2
  from sidestep_engine.models.loader import load_decoder_for_training
  from sidestep_engine.vendor.loha_utils import inject_loha_into_dit


  def main() -> None:
      ap = argparse.ArgumentParser()
      ap.add_argument("--base-ckpt-dir", required=True,
                      help="Parent dir containing acestep-v15-xl-base/ subdir")
      ap.add_argument("--variant", default="xl-base")
      ap.add_argument("--adapter-dir", required=True,
                      help="Side-Step run dir (with sidestep_adapter_config.json + final/loha_weights.safetensors)")
      ap.add_argument("--out", required=True)
      ap.add_argument("--device", default="cpu", help="cpu or cuda:0 (use cpu if GPU busy)")
      ap.add_argument("--weight", type=float, default=1.0,
                      help="Adapter merge strength (1.0 = full)")
      ap.add_argument("--checkpoint-name", default="final",
                      help="Subfolder under adapter-dir containing loha_weights.safetensors")
      args = ap.parse_args()

      base_ckpt_dir = Path(args.base_ckpt_dir).resolve()
      adapter_dir = Path(args.adapter_dir).resolve()
      out_dir = Path(args.out).resolve()
      base_model_dir = base_ckpt_dir / f"acestep-v15-{args.variant}"

      cfg_path = adapter_dir / "sidestep_adapter_config.json"
      weights_path = adapter_dir / args.checkpoint_name / "loha_weights.safetensors"
      if not cfg_path.exists():
          sys.exit(f"[FAIL] Config not found: {cfg_path}")
      if not weights_path.exists():
          sys.exit(f"[FAIL] Weights not found: {weights_path}")

      print(f"[1/6] Loading base model {args.variant} on {args.device}")
      model = load_decoder_for_training(
          checkpoint_dir=str(base_ckpt_dir),
          variant=args.variant,
          device=args.device,
          precision="fp32",
      )

      print(f"[2/6] Loading LoHA config: {cfg_path}")
      loha_config = LoHAConfigV2.from_json(cfg_path)
      print(f"      dim={loha_config.linear_dim}, alpha={loha_config.linear_alpha}, "
            f"target_modules={loha_config.target_modules}")

      print(f"[3/6] Injecting LoHA modules + loading trained weights via load_weights() (mirrors API path)")
      model, lycoris_net, info = inject_loha_into_dit(model, loha_config)
      print(f"      LyCORIS modules created: {len(lycoris_net.loras)}")
      load_state = lycoris_net.load_weights(str(weights_path))
      if load_state.get("missing keys"):
          print(f"      [WARN] Missing keys (sample): {load_state['missing keys'][:3]}")
      if load_state.get("unexpected keys"):
          print(f"      [WARN] Unexpected keys (sample): {load_state['unexpected keys'][:3]}")

      print(f"[4/6] Manually baking diff into base (uses get_weight() directly, mirroring forward pass)")
      # NOTE: forward pass (loha.py:311) uses self.get_weight() * self.scalar (NOT get_diff_weight).
      # get_diff_weight() double-scales: it multiplies get_weight() by self.scale, but get_weight()
      # already applies scale via gamma=scale. We must use get_weight() directly to match forward.
      with torch.no_grad():
          merged_count = 0
          skipped_count = 0
          for lora_module in lycoris_net.loras:
              base_module = lora_module.org_module[0]
              try:
                  diff_weight = lora_module.get_weight(base_module.weight.shape)
              except Exception as exc:
                  print(f"      [SKIP] {lora_module.lora_name}: get_weight failed ({exc})")
                  skipped_count += 1
                  continue
              if diff_weight is None:
                  skipped_count += 1
                  continue
              diff_weight = diff_weight.reshape(base_module.weight.shape).to(
                  base_module.weight.device, base_module.weight.dtype
              )
              scalar = float(getattr(lora_module, "scalar", 1.0))
              multiplier = float(getattr(lora_module, "multiplier", 1.0))
              base_module.weight.data.add_(diff_weight * (args.weight * scalar * multiplier))
              merged_count += 1
          print(f"      Merged {merged_count} modules, skipped {skipped_count}")

      print(f"[5/6] Removing lycoris_net so save_pretrained doesn't serialize ignored adapter weights")
      lycoris_net.restore()
      if hasattr(model, "decoder"):
          for attr in [a for a in dir(model.decoder) if a.startswith("_lycoris")]:
              try:
                  delattr(model.decoder, attr)
                  print(f"      removed model.decoder.{attr}")
              except Exception:
                  pass

      print(f"[6/6] Saving merged model to {out_dir}")
      out_dir.mkdir(parents=True, exist_ok=True)
      model.save_pretrained(str(out_dir), safe_serialization=True)

      for src in base_model_dir.iterdir():
          if src.is_file() and (
              src.suffix in (".py", ".md", ".json", ".txt")
              or src.name == "silence_latent.pt"
          ):
              dst = out_dir / src.name
              if not dst.exists() or src.name in ("config.json", "silence_latent.pt"):
                  shutil.copy2(src, dst)

      print(f"[OK] Done. Merged model at: {out_dir}")
      print(f"     Load with: variant=\"{args.variant}\", model_path=\"{out_dir}\"")


  if __name__ == "__main__":
      main()
  ```
  </details>

plz tag me #34 - Looking forward for the full fine-tune.
